src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx

```diff
@@ -0,0 +1,467 @@
+  ) => Promise<void>;
+  http: HttpSetup;
+  workspaces: WorkspacesStart;
+  duplicateMode: DuplicateMode;
+  notifications: NotificationsStart;
+  selectedSavedObjects: SavedObjectWithMetadata[];
+}
+
+interface Props extends ShowDuplicateModalProps {
+  onClose: () => void;
+}
+
+interface State {
+  allSelectedObjects: SavedObjectWithMetadata[];
+  workspaceOptions: WorkspaceOption[];
+  allWorkspaceOptions: WorkspaceOption[];
+  targetWorkspaceOption: WorkspaceOption[];
+  isLoading: boolean;
+  isIncludeReferencesDeepChecked: boolean;
+  savedObjectTypeInfoMap: Map<string, [number, boolean]>;
+}
+
+function capitalizeFirstLetter(str: string): string {
+  return str.charAt(0).toUpperCase() + str.slice(1);
+}
+
+export class SavedObjectsDuplicateModal extends React.Component<Props, State> {
+  private isMounted = false;
+
+  constructor(props: Props) {
+    super(props);
+
+    this.state = {
+      allSelectedObjects: this.props.selectedSavedObjects,
+      workspaceOptions: [],
+      allWorkspaceOptions: [],
+      targetWorkspaceOption: [],
+      isLoading: false,
+      isIncludeReferencesDeepChecked: true,
+      savedObjectTypeInfoMap: new Map<string, [number, boolean]>(),
+    };
+  }
+
+  workspaceToOption = (
+    workspace: WorkspaceAttribute,
+    currentWorkspaceName?: string
+  ): WorkspaceOption => {
+    // add (current) after current workspace name
+    let workspaceName = workspace.name;
+    if (workspace.name === currentWorkspaceName) {
+      workspaceName += ' (current)';
+    }
+    return {
+      label: workspaceName,
+      key: workspace.id,
+      value: workspace,
+    };
+  };
+
+  async componentDidMount() {
+    const { workspaces } = this.props;
+    const currentWorkspace = workspaces.currentWorkspace$.value;
+    const currentWorkspaceName = currentWorkspace?.name;
+    const targetWorkspaces = this.getTargetWorkspaces();
+
+    // current workspace is the first option
+    const workspaceOptions = [
+      ...(currentWorkspace ? [this.workspaceToOption(currentWorkspace, currentWorkspaceName)] : []),
+      ...targetWorkspaces
+        .filter((workspace: WorkspaceAttribute) => workspace.name !== currentWorkspaceName)
+        .map((workspace: WorkspaceAttribute) =>
+          this.workspaceToOption(workspace, currentWorkspaceName)
+        ),
+    ];
+
+    this.setState({
+      workspaceOptions,
+      allWorkspaceOptions: workspaceOptions,
+    });
+
+    const { duplicateMode } = this.props;
+    if (duplicateMode === DuplicateMode.All) {
+      const { allSelectedObjects } = this.state;
+      const categorizedObjects = groupBy(allSelectedObjects, (object) => object.type);
+      const savedObjectTypeInfoMap = new Map<string, [number, boolean]>();
+      for (const [savedObjectType, savedObjects] of Object.entries(categorizedObjects)) {
+        savedObjectTypeInfoMap.set(savedObjectType, [savedObjects.length, true]);
+      }
+      this.setState({ savedObjectTypeInfoMap });
+    }
```

Comment:
Seems the code should be move to constructor or beforeMount lifeCycle.

Reply:
I will move the logic of obtaining workspace options to the `constructor()`. The logic of obtaining all saved objects will still be placed in `componentDidMount()`, otherwise the objects will not be obtained.